### PR TITLE
[fud] Add save_temps option to Xilinx synthesis

### PR DESCRIPTION
**Pull request by sampsyo**
This change builds on #850 (and depends on it, which is why this is a draft for now and is not worth reviewing until that's merged).

It's a pretty small additional feature: it adds support for the `save_temps` config option in the Xilinx bitstream compilation stage, which is called `xclbin`. This option works in both SSH and local mode, and it just keeps around the temporary directory that would otherwise be deleted at the end of the fud execution. This should make it super easy to check things out manually when something goes wrong.

I added a simple utility class, `FreshDir`, that creates a new empty directory in cwd. I hope this will be useful for other vendor-toolchain-related fud stages that need sandbox directories.

There is new documentation in the `synthesis.md` chapter for how to use the new flag.

### CODE REVIEW

**Comment by sampsyo** [comment]
Rebased and ready to look at!

File: fud/fud/stages/remote_context.py [hunk 1]
```diff
@@ -120,18 +120,20 @@ def run_remote(client: SourceType.UnTyped, tmpdir: SourceType.String):
 
         run_remote(client, tmpdir)
 
-    def _close(self, client, remote_tmpdir):
+    def _close(self, client, remote_tmpdir, keep=False):
```
Comment (cgyurgyik) [inline fud/fud/stages/remote_context.py:123]: Nit: prefer a more verbose name, e.g. `keep_tmpdir` or `keep_remote_tmpdir`.

File: fud/fud/stages/xilinx/xclbin.py [hunk 2]
```diff
@@ -200,14 +205,14 @@ def read_file(
...
 
         if self.remote_exec.use_ssh:
-            xclbin = self.remote_exec.close_and_get(
+            return self.remote_exec.close_and_get(
                 client,
                 tmpdir,
                 "xclbin/kernel.xclbin",
+                keep=self.save_temps,
             )
         else:
```
Comment (cgyurgyik) [inline fud/fud/stages/xilinx/xclbin.py:214]: Super nit: `else` not really needed here, but understand some people like the symmetry.

File: fud/fud/stages/xilinx/xclbin.py [hunk 3]
```diff
@@ -40,6 +40,8 @@ def __init__(self, config):
         self.remote_exec = RemoteExecution(self)
         self.temp_location = self.config["stages", self.name, "temp_location"]
 
+        self.save_temps = bool(self.config["stages", self.name, "save_temps"])
```
Comment (cgyurgyik) [inline fud/fud/stages/xilinx/xclbin.py:43]: Maybe a comment explaining why this is added would prove useful for future readers. It wasn't obvious to me until I read your PR description.

**Comment by rachitnigam** [comment]
LGTM! Chris's nits are a superset of mine so address and merge!

File: fud/fud/stages/remote_context.py (hunk 1 above)
Comment (sampsyo) [reply fud/fud/stages/remote_context.py]: Sounds good; changed!

File: fud/fud/stages/xilinx/xclbin.py (hunk 3 above)
Comment (sampsyo) [reply fud/fud/stages/xilinx/xclbin.py:45]: Good call; added a short comment.

**Comment by sampsyo** [comment]
Thanks, y'all! Merging now.

**Pull request merged**